README.md
# NetUtil

## statement

NetUtil is a compact and efficient Java library for sending and receiving messages using the OpenSoundControl (OSC) protocol. It is (C)opyright 2004&ndash;2013 by Hanns Holger Rutz. All rights reserved. NetUtil is released under the [GNU Lesser General Public License](licenses/NetUtil-License.txt) and comes with absolutely no warranties. To contact the author, send an email to `contact at sciss.de`.

For project status, API and current version, visit [github.com/Sciss/NetUtil](https://github.com/Sciss/NetUtil).

Note that this project is not actively developed any more (unless there are bugs to fix). For a current OSC library in the Scala programming language, please visit [github.com/Sciss/ScalaOSC](https://github.com/Sciss/ScalaOSC).

## requirements / building

NetUtil requires Java 1.4+. NetUtil has also been successfuly running on Android. It uses [sbt](http://sbt.github.com/) 0.12 for building. You can use the included `sbt` shell script if you do not want to install sbt.

To compile `sbt compile`, to generate a jar `sbt package`.

## linking

To use NetUtil in your project, you can link to the following [Maven](http://search.maven.org) artifact:

    GroupId: de.sciss
    ArtifactId: netutil
    Version: 1.0.0

## documentation

Documentation comes in the form of JavaDoc. The generate the docs, run `sbt doc`. The resulting file is in `target/api/index.html`.

You can run some demos using `sbt`. First run `./sbt`. At the prompt:

    > run

This will print to available options. E.g.

    > run --testPingPong

## noteworthy links

- [www.opensoundcontrol.org](http://www.opensoundcontrol.org) &ndash; information about OSC specifications and implementations
- [Illposed JavaOSC](http://www.illposed.com/software/javaosc.html) &ndash; another OSC library for Java
- [oscP5](http://www.sojamo.de/iv/index.php?n=11) &ndash; OSC library for Processing
- [flosc](http://www.benchun.net/flosc/) &ndash; OSC library (written in Java) for bridging OSC clients and Macromedia Flash
- [jmDNS](http://jmdns.sourceforge.net/) &ndash; automatic service discovery library for Java.

Here is some sporadic list of projects which seem to use NetUtil. if you want to have your project added here, send me the links:

- [Androidome](http://code.google.com/p/androidome/) &ndash; android-based emulator for the monome music making device
- [jReality](http://www3.math.tu-berlin.de/jreality/)
- [OscVstBridge](http://www.savedbytechnology.com/main4/oscvstbridge.htm) &ndash; VST plug-in
- [Frozen Bubble OSCified](http://blog.cappel-nord.de/2008/07/frozen-bubble/)
- [DiABlu](http://diablu.jorgecardoso.eu/) &ndash; Scout and LegOSC
- [AMICO](http://amico.sourceforge.net/")

## to-do / known issues

- add special client and server listeners so one can detect channels/connections opening and closing
- add helper classes (multi-server wrapper, message deferrer)



CHANGES.md
# Change History

In the course of the development of NetUtil, some API changes were made. Generally, changing the API is considered bad because it breaks previous code (it is not backward compatible). However, I tried to minimize these changes. Generally, changes have only been made, when the API proved suboptimal or new features were added which would introduce ugly API when enforcing backward compatibility.

This document tries to help updating code that uses old versions of NetUtil.

## v0.39 (jun 2011 - SVN rev. 23)</h4>

- OSCClient and OSCTransmitter creation: socket's IP defaults to "0.0.0.0" now instead of InetAddress.getLocalHost(). Allegedly that was responsible for an issue on Android.

## v0.38 (may 2009 - SVN rev. 21)

- Removed makedoc.sh in favour of a doc ant target

## v0.37 (may 2009 - SVN rev. 18)

- OSCServer creation: socket's IP defaults to "0.0.0.0" now instead of InetAddress.getLocalHost(); added tests

## v0.36 (may 2009 - SVN rev. 16)

- Fixed bug in OSCReceiver.newUsing( DatagramChannel )
- Increased OSCPacketCodec.Atom visibility
- Fixed confusing error message with null OSC message arguments
- Fixed typo: OSCBundle.setTimeTagRow becomes setTimeTagRaw
- OSCReceiver creation: socket's IP defaults to "0.0.0.0" now instead of InetAddress.getLocalHost(), since this way messages both from localhost and loopback sockets are received.

## v0.35 (apr 2008 - SVN rev. 6)

- Improved and fixed javadoc comments for OSCTransmitter, OSCReceiver, OSCClient, OSCServer

## v0.34 (apr 2008 - SVN rev. 5)

- Added javadoc comments for OSCPacketCodec

## v0.33 (jul 2007)

- Added separate OSCPacketCodec class and factory methods for OSCClient, OSCServer, OSCTransmitter, OSCReceiver, using this new codec. The codec allows to specify 64-bit primitives behaviour among others. The SpecificOSCMessage class was dropped.
- String ('s' type) argument decoding now follows the given codec. Default codec uses UTF-8 on all platforms.
- Bug fix in OSCTransmitter.newUsing( String, int ).
- Slightly increased decoding performance.

## v0.32 (mar 2007)

- String ('s' type) argument decoding correctly uses platform default charset for ascii characters >127.
- using Eclipse and Ant now.

## v0.31 (oct 2006)

- Allows &quot;revivable&quot; channels and clients. Improved TCP support.

## v0.30 (oct 2006)

- Integration of TCP mode as alternative to UDP.
- This suggested API changes (see separate section at the end of this document).
- New classes <code>OSCClient</code> and <code>OSCServer</code> simplify the use of receivers and transmitters.
- Improved javadocs

## v0.26 (jul 2006)

- Minor improvements in OSCReceiver

## v0.25 (nov 2005)

- Bug fixes : decoding blob arguments

## v0.24 (sep 2005)

- API change: the <code>OSCListener</code> interface is modified to include the timetag of the incoming message.
- Bug fixes : byte boundary alignment of blobs, packet size calculation with blobs and nested messages.
- Small improvements (hexdump looks better + receiver thread catches runtime exceptions)

## v0.22 (sep 2005)

- OSCMessage allows byte[] arguments which are encoded as a blob.

## v0.2 (aug 2005)

- <code>OSCReceiver</code> is accompanied by a new <code>OSCTransmitter</code> class which is really simple but saves you from typing the same five lines over and over again. Their common super class is <code>AbstractOSCCommunicator</code> which provides a <code>dumpOSC</code> method.
- It turned out that binding an unbound channel in the <code>OSCReceiver</code> constructor to <code>&quot;127.0.0.1&quot;</code> would silently disallow that channel to send messages outside the local computer. The behaviour has therefore changed to bind the channel to the loopback address only if the filter address is a loopback address (indicating usage on the local machine only), otherwise the local IP address is used.

## v0.14 (aug 2005)

- OSCReceiver has a new contract of not allowing channels to be connected. This allows you to use the channel both for sending and receiving messages. Print messages have been added to OSCPacket and a debugDump method to OSCReceiver

## v0.12 (jul 2005)

- tries to avoid a runtime error discovered on linux (sun jdk 1.5)

## v0.11 (jun 2005)

- first version, an extract from meloncillo basically

# API Changes

The following sections only highlight backward incompatible API changes.

## API changes from v0.22 to v0.24

 - in interface de.sciss.net.OSCListener
  old signature:

			public void messageReceived( OSCMessage m, SocketAddress addr );

  new signature:

			public void messageReceived( OSCMessage m, SocketAddress addr, long when );

  this change was necessary to pass bundle execution times to the listener. to update
  old code, simply replace the old signatures with the new ones and ignore the
  bundle time (long when).

## API changes from v0.26 to v0.30

- these changes were made in the course of TCP integration
	- deletion of class de.sciss.net.AbstractOSCCommunicator
		why? the class was too specific about the network channel type. the class name
		is ugly. the class was used to define constants, which is generally done
		by using an interface.

		therefore, the interface de.sciss.net.OSCChannel was created which includes
		some of the methods of AbstractOSCCommunicator.

		in your code, try to replace references to AbstractOSCCommunicator by their
		concrete subclass OSCReceiver or OSCTransmitter. Getting the DatagramChannel
		is impossible now. Instead memorize the channel you pass to one of the
		constructors in OSCReceiver or OSCTransmitter. If you used the empty
		constructor, create an explicit channel that can be memorized.

		if you just need to learn the local socket address, you can call
		getLocalAddress() now.

	- de.sciss.net.OSCReceiver and de.sciss.net.OSCTransmitter became abstract
		making these classes abstract provides a much more beautiful way of dealing
		with different transport types than using internal delegates.


trunk/src/de/sciss/net/OSCBundle.java
/*
 *  OSCBundle.java
 *  de.sciss.net (NetUtil)
 *
 *  Copyright (c) 2004-2008 Hanns Holger Rutz. All rights reserved.
 *
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 *	For further information, please contact Hanns Holger Rutz at
 *	contact@sciss.de
 *
 *
 *  Changelog:
 *		25-Jan-05	created from de.sciss.meloncillo.net.OSCBundle
 *		01-Mar-05	sample accurate timing through new setTimeTag() method
 *					(annot. 12-May-05 : scsynth OSC processing is *not* audiorate accurate)
 *		26-May-05	moved to de.sciss.net package ; slight modifications
 *		21-Jun-05	extended javadoc
 *		28-Apr-07	moved codec stuff to OSCPacketCodec
 */

package de.sciss.net;

import java.util.ArrayList;
import java.util.List;
//import java.util.Map;

/**
 *  Implementation of the OSC-Bundle
 *  which assembles several OSC-Packets
 *  under the same timetag. See <A HREF="http://opensoundcontrol.org/spec-1_0">opensoundcontrol.org/spec-1_0</A> for
 *	the specification of the bundle format.
 *	<p>
 *	The bundle time can be calculated in
 *	different ways: specifying a <code>long</code>
 *	which represents the milliseconds since
 *	1 jan 1970 as returned by <code>System.currentTimeMillis()</code>,
 *	produces the required network time tag
 *	as required by the OSC specification.
 *	Alternatively, using a <code>float</code> 
 *	for a relative time offset in seconds can
 *	be used when running SuperCollider in offline
 *	mode. Third, there is a version that calculates
 *	a sample accurate time tag. However, scsynth
 *	doesn't process bundles with this accuracy, so
 *	it's kind of useless.
 *	<p>
 *	To assemble a bundle, you create a new instance of
 *	<code>OSCBundle</code>, call <code>addPacket</code>
 *	one or several times, then write the contents of
 *	the bundle to a <code>ByteBuffer</code> using the
 *	method <code>encode</code>. The byte buffer can then
 *	be written to a <code>DatagramChannel</code>. Here is an example:
 *	<pre>
 *      OSCBundle       bndl;
 *      DatagramChannel dch = null;
 *
 *      final ByteBuffer buf        = ByteBuffer.allocateDirect( 1024 );
 *      final SocketAddress addr    = new InetSocketAddress( "localhost", 57110 );
 *      final long serverLatency    = 50;
 *      final long now              = System.currentTimeMillis() + serverLatency;
 *
 *      try {
 *          dch = DatagramChannel.open();
 *          dch.configureBlocking( true );
 *          bndl = new OSCBundle( now );
 *          bndl.addPacket( new OSCMessage( "/s_new", new Object[] { "default",
 *                          new Integer( 1001 ), new Integer( 1 ), new Integer( 0 ),
 *                          "out", new Integer( 0 ), "freq", new Float( 666 ), "amp", new Float( 0.1f )}));
 *          bndl.encode( buf );
 *          buf.flip();
 *          dch.send( buf, addr );
 *          buf.clear();
 *          bndl = new OSCBundle( now + 2000 ); // two seconds later
 *          bndl.addPacket( new OSCMessage( "/n_free", new Object[] { new Integer( 1001 )}));
 *          bndl.encode( buf );
 *          buf.flip();
 *          dch.send( buf, addr );
 *      }
 *      catch( IOException e2 ) {
 *          System.err.println( e2.getLocalizedMessage() );
 *      }
 *      finally {
 *          if( dch != null ) {
 *              try {
 *                  dch.close();
 *              }
 *              catch( IOException e4 ) {};
 *          }
 *      }
 *	</pre>
 *
 *	Note that this example uses the old way of sending messages.
 *	A easier way is to create an <code>OSCTransmitter</code> which
 *	handles the byte buffer for you. See the <code>OSCReceiver</code> doc
 *	for an example using a dedicated transmitter.
 *
 *  @author		Hanns Holger Rutz
 *  @version	0.33, 28-Apr-07
 *
 *	@see		OSCReceiver
 */
public class OSCBundle
extends OSCPacket
{
	/**
	 *  This is the initial string
	 *  of an OSC bundle datagram
	 */
	public static final String TAG   = "#bundle";

	/**
	 *  The special timetag value
	 *  to indicate that the bundle be
	 *	processed as soon as possible
	 */
	public static final long NOW   = 1;

	private static final long SECONDS_FROM_1900_TO_1970 = 2208988800L;

//	private static final byte[] cmd1  = { 0x23, 0x62, 0x75, 0x6E, 0x64, 0x6C, 0x65, 0x00 }; // "#bundle" (4-aligned)
	private long timetag;   // 64 bit fixed point seconds since 1 jan 1900
	protected final List collPackets  = new ArrayList();

	/**
	 *  Creates a new empty OSCBundle
	 *  with timetag set to &quot;immediately&quot;.
	 *	SuperCollider recognizes this special timetime
	 *	to process the bundle just when it arrives.
	 */
	public OSCBundle()
	{
		super();

		timetag = NOW;	// special code
	}

	/**
	 *  Creates a new empty OSCBundle
	 *  with timetag specified by 'when'
	 *  which is milliseconds since 1 jan 1970
	 *  as returned by System.currentTimeMillis().
	 *  This is converted into an absolute time
	 *  offset since 1 jan 1900 as required by
	 *  the OSC specs.
	 *
	 *  @param  when	absolute time tag for the bundle
	 *  @see	java.lang.System#currentTimeMillis()
	 */
	public OSCBundle( long when )
	{
		super();
		setTimeTagAbsMillis( when );
	}

	/**
	 *  Creates a new empty OSCBundle
	 *  with timetag specified by 'when'
	 *  which is seconds relative to start
	 *  of session. This relative time offset
	 *  with origin of zero is understood by
	 *  SuperCollider in NonRealTime mode.
	 *  (see example in Non-Realtime-Synthesis.rtf).
	 *
	 *  @param  when	relative time tag for the bundle
	 */
	public OSCBundle( double when )
	{
		super();
		setTimeTagRelSecs( when );
	}

	/**
	 *  Creates a new empty OSCBundle
	 *  with timetag specified by a sample
	 *	frame offset and an absolute time
	 *	in milliseconds since 1 jan 1970
	 *  as returned by <code>System.currentTimeMillis()</code>
	 *
	 *	@param	absMillisOffset	time offset as returned by <code>System.currentTimeMillis</code>
	 *	@param	sampleFrames	this offset is added to the milli second offset.
	 *	@param	sampleRate		used in conjunction with <code>sampleFrames</code> to
	 *							calculate the time offset.
	 */
	public OSCBundle( long absMillisOffset, long sampleFrames, int sampleRate )
	{
		super();
		setTimeTagSamples( absMillisOffset, sampleFrames, sampleRate );
	}
	
	/**
	 *  Adds a new <code>OSCPacket</code> to the tail
	 *  of the bundle. Passing <code>null</code> is
	 *  allowed in which case no action
	 *  is performed.
	 *
	 *  @param  p   the packet to add to the tail of the bundle
	 */
	public void addPacket( OSCPacket p )
	{
		if( p != null ) {
			synchronized( collPackets ) {
				collPackets.add( p );
			}
		}
	}

	/**
	 *  Gets the <code>OSCPacket</code> at the provided
	 *  index which must be between 0 inclusive
	 *  and <code>getPacketCount()</code> exclusive.
	 *  If bundles are nested, each nested
	 *  bundle will count as one packet of course.
	 *
	 *  @param  idx		index of the packet to get
	 *  @return packet at index <code>idx</code>
	 */
	public OSCPacket getPacket( int idx )
	{
		synchronized( collPackets ) {
			return (OSCPacket) collPackets.get( idx );
		}
	}

	/**
	 *  Returns the number of packets
	 *  currently assembled in this bundle.
	 *  If bundles are nested, each nested
	 *  bundle will count as one packet of course.
	 *
	 *  @return number of packets assembled in this bundle
	 */
	public int getPacketCount()
	{
		synchronized( collPackets ) {
			return collPackets.size();
		}
	}

	/**
	 *  Removes the specified packet
	 *
	 *  @param  idx the index of the packet to remove
	 */
	public void removePacket( int idx )
	{
		synchronized( collPackets ) {
			collPackets.remove( idx );
		}
	}

//	public int getSize()
//	throws IOException
//	{
//		synchronized( collPackets ) {
//			int result = cmd1.length + 8 + (collPackets.size() << 2); // name, timetag, size of each bundle element
//
//			for( int i = 0; i < collPackets.size(); i++ ) {
//				result += ((OSCPacket) collPackets.get( i )).getSize();
//			}
//
//			return result;
//		}
//	}

	/**
	 *  Sets the bundle's timetag
	 *  specified by a long
	 *  which is milliseconds since 1 jan 1970
	 *  as returned by <code>System.currentTimeMillis()</code>.
	 *  This is converted into an absolute time
	 *  offset since 1 jan 1900 as required by
	 *  the OSC specs.
	 *
	 *  @param  when	absolute time tag for the bundle
	 *  @see	java.lang.System#currentTimeMillis()
	 */
	public void setTimeTagAbsMillis( long when )
	{
		final long secsSince1900	= when / 1000 + SECONDS_FROM_1900_TO_1970;
		final long secsFractional	= ((when % 1000) << 32) / 1000;
		timetag						= (secsSince1900 << 32) | secsFractional;
	}
	
	public void setTimeTagRaw( long raw )
	{
		timetag = raw;
	}

	/**
	 *  Sets the bundle's timetag
	 *  specified by a double
	 *  which is seconds relative to start
	 *  of session. This relative time offset
	 *  with origin of zero is understood by
	 *  SuperCollider in NonRealTime mode.
	 *  (see example in Non-Realtime-Synthesis.rtf).
	 *
	 *  @param  when	relative time tag for the bundle
	 */
	public void setTimeTagRelSecs( double when )
	{
//		timetag	= ((long) when << 32) | (long) ((when % 1.0) * 0x100000000L + 0.5);
		timetag	= ((long) when << 32) + (long) ((when % 1.0) * 0x100000000L + 0.5);
	}

	/**
	 *  Sets the bundle's timetag as
	 *	a combination of system absolute time
	 *	and sample offset. note that
	 *	this is not too useful, because supercollider
	 *	will execute OSC bundles not with
	 *	audiorate but controlrate precision!!
	 *
	 *	@param	absMillisOffset	time offset as returned by <code>System.currentTimeMillis</code>
	 *	@param	sampleFrames	this offset is added to the milli second offset.
	 *	@param	sampleRate		used in conjunction with <code>sampleFrames</code> to
	 *							calculate the time offset.
	 */
	public void setTimeTagSamples( long absMillisOffset, long sampleFrames, int sampleRate )
	{
		final double seconds= (double) sampleFrames / (double) sampleRate + (double) absMillisOffset / 1000;
		timetag				= (((long) seconds + SECONDS_FROM_1900_TO_1970) << 32) + (long) ((seconds % 1.0) * 0x100000000L + 0.5);
	}

	/**
	 *  Returns the raw format time tag
	 *  of the bundle
	 *
	 *  @return  the bundle's timetag in OSC format
	 *
	 *	@todo	a utility method to convert this
	 *			to a more useful value
	 */
	public long getTimeTag()
	{
		return timetag;
	}
}